Use runuser in non-su-compatible mode

lukebakken opened the pull request:
Fixes #44 

Inspired by #40 where I didn't have the `runuser` arguments just right, then by ab705f8359fada8cc9ccdc72ff638657956f2d06 and aa84ddcf36b67e0196b5cfa81bb34008bdf70a43.

Also, remove argument quoting which is not necessary as long as original script arguments are passed to `/sbin/runuser` and `/bin/su` as arguments instead of one big string.

Tested using the following scripts -

Save this to `/var/lib/rabbitmq/runuser-test` and make it executable:

```sh
#!/bin/sh
echo subcmd via runuser:
/sbin/runuser -u rabbitmq -- /var/lib/rabbitmq/runuser-subcmd "$@"
echo
echo subcmd via su:
/bin/su -s /bin/sh rabbitmq -- /var/lib/rabbitmq/runuser-subcmd "$@"
```

Save this to `/var/lib/rabbitmq/runuser-subcmd` and make it executable:

```sh
#!/bin/sh
echo "$0"
echo "ID: $(id)"
echo "ARGS: $@"
for ARG in "$@"
do
    echo "ARG: $ARG"
done
```

Example command and output from my CentOS 7 VM:

```
[root@localhost ~]# /var/lib/rabbitmq/runuser-test ARG0 'ARG 1' '"ARG 2"' "ARG 3"                                                                                                                                                                                               
subcmd via runuser:                                                                                                                                                                                                                                                             
/var/lib/rabbitmq/runuser-subcmd                                                                                                                                                                                                                                                
ID: uid=996(rabbitmq) gid=994(rabbitmq) groups=994(rabbitmq) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023                                                                                                                                                      
ARGS: ARG0 ARG 1 "ARG 2" ARG 3                                                                                                                                                                                                                                                  
ARG: ARG0                                                                                                                                                                                                                                                                       
ARG: ARG 1                                                                                                                                                                                                                                                                      
ARG: "ARG 2"                                                                                                                                                                                                                                                                    
ARG: ARG 3                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                                                
subcmd via su:                                                                                                                                                                                                                                                                  
/var/lib/rabbitmq/runuser-subcmd                                                                                                                                                                                                                                                
ID: uid=996(rabbitmq) gid=994(rabbitmq) groups=994(rabbitmq) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
ARGS: ARG0 ARG 1 "ARG 2" ARG 3
ARG: ARG0
ARG: ARG 1
ARG: "ARG 2"
ARG: ARG 3
```

lukebakken committed:
Use runuser in non-su-compatible mode

Remove argument quoting which is not necessary as long as command is
passed to `/sbin/runuser` and `/bin/su` as arguments instead of a
string.

Fixes #44.
[#150221349]

hairyhum commented:
`-u` argument is not supported in some distributions.
https://linux.die.net/man/1/runuser

lukebakken commented:
@hairyhum do you know which distros use the `coreutils` version vs the `util-linux` one?

lukebakken commented:
This is interesting, neither `su` nor `runuser` are part of [`coreutils`](http://git.savannah.gnu.org/gitweb/?p=coreutils.git;a=tree;f=src;h=164741dace45d3598a58a4c1328b3578d94e68ad;hb=HEAD) anymore. I'll see when they were removed and will correlate with our supported distro lists.

hairyhum commented:
@lukebakken you can search failures in `test-package-` jobs for 3.7 pipelines.

lukebakken committed:
Detect util-linux version of runuser

Alternative is to always use runuser invocation in else clause.

Fixes #44.
[#150221349]

lukebakken commented:
I modified `runuser-test` to the following:

```
#!/bin/sh
echo subcmd via runuser:
if /sbin/runuser --version | grep -qF util-linux
then
    /sbin/runuser -u rabbitmq -- /var/lib/rabbitmq/runuser-subcmd "$@"
else
    /sbin/runuser -s /bin/sh -- rabbitmq /var/lib/rabbitmq/runuser-subcmd "$@"
fi
echo "runuser exit code: $?"
echo
echo subcmd via su:
/bin/su -s /bin/sh rabbitmq -- /var/lib/rabbitmq/runuser-subcmd "$@"
echo "su exit code: $?"
```

Yes, this is more complicated but I added a `TODO` in the code to eventually simplify the script. The alternative is to just use the `/sbin/runuser` invocation in the `else` clause.

dumbbell commented on scripts/rabbitmq-script-wrapper:
```diff
@@ -83,10 +77,21 @@ exec_script_as_rabbitmq() {
+    # removed
+    if /sbin/runuser --version | grep -qF util-linux
+    then
+        exec /sbin/runuser -u rabbitmq -- "/usr/lib/rabbitmq/bin/$SCRIPT" "$@"
```
`"$@"` is always empty here because `exec_script_as_root`, called on line 35, isn't passed `"$@"`. Therefore any use of `rabbitmqctl` as root show the usage and exits with 64.

I will commit the fix I just tried. Everything is working as expected after that.

lukebakken replied:
@dumbbell thanks for catching that ... I forgot to run `shellcheck` which would have found that:

```
$ shellcheck scripts/rabbitmq-script-wrapper

In scripts/rabbitmq-script-wrapper line 35:
    exec_script_as_root
    ^-- SC2119: Use exec_script_as_root "$@" if function's $1 should mean script's $1.


In scripts/rabbitmq-script-wrapper line 77:
exec_script_as_root() {
^-- SC2120: exec_script_as_root references arguments, but none are ever passed.
```

dumbbell commented:
Before merging, I will also rebase the commits to reference the GitHub issues and Pivotal Tracker story.
